File: csharp/beer-song/BeerSong.cs
using System;
using System.Text;

public static class BeerSong
{
    public static string Recite(int startBottles, int takeDown)
    {
        var recited = new StringBuilder();
        
        string plural(int count) => count == 1 ? "" : "s";
        string number(int count, bool capitalise = false) => count == 0 ? $"{(capitalise ? "N" : "n")}o more" : count.ToString();
        string onTheWall(int count, bool capitalise = false) => $"{number(count, capitalise)} bottle{plural(count)} of beer on the wall";
        string ofBeer(int count) => $"{number(count)} bottle{plural(count)} of beer.";

        while(takeDown > 0)
        {
            recited.Append($"{onTheWall(startBottles, true)}, {ofBeer(startBottles)}\n");

            startBottles -= 1;
            takeDown -= 1;

            if(startBottles >= 0)
                recited.Append($"Take {(startBottles > 0 ? "one" : "it")} down and pass it around, {onTheWall(startBottles)}.");
            else
                recited.Append($"Go to the store and buy some more, {onTheWall(99)}.");

            if(takeDown > 0)
                recited.Append("\n\n");
        }

        return recited.ToString();
    }
}

File: fsharp/beer-song/BeerSong.fs
module BeerSong

let recite startBottles toTakeDown = 

    let rec reciteWithTailRecursion prior bottles takeDown = 
        let bottleLabel line = 
            match line with
            | 0 -> "No more bottles"
            | 1 -> "1 bottle"
            | _ -> sprintf "%d bottles" line

        let currentLabel = bottleLabel bottles
        let nextLabel = bottleLabel (bottles - 1)

        let lines = [
            sprintf "%s of beer on the wall, %s of beer." currentLabel (currentLabel.ToLower());
            (match bottles with
            | 0 -> "Go to the store and buy some more, 99 bottles of beer on the wall."
            | _ -> sprintf "Take %s down and pass it around, %s of beer on the wall." (if bottles = 1 then "it" else "one") (nextLabel.ToLower()))
        ]

        match takeDown with
        | _ when takeDown <= 1 -> [prior @ lines]
        | _ -> reciteWithTailRecursion (prior @ lines @ [""]) (bottles - 1) (takeDown - 1)

    reciteWithTailRecursion [] startBottles toTakeDown |> List.concat